pkg/scheduler/core/division_algorithm.go

```diff
@@ -0,0 +1,205 @@
+package core
+
+import (
+	"fmt"
+	"sort"
+
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+// TargetClustersList is a slice of TargetCluster that implements sort.Interface to sort by Value.
+type TargetClustersList []workv1alpha2.TargetCluster
+
+func (a TargetClustersList) Len() int           { return len(a) }
+func (a TargetClustersList) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a TargetClustersList) Less(i, j int) bool { return a[i].Replicas > a[j].Replicas }
+
+// divideReplicasByDynamicWeight assigns a total number of replicas to the selected clusters by the dynamic weight list.
+func divideReplicasByDynamicWeight(clusters []*clusterv1alpha1.Cluster, dynamicWeight policyv1alpha1.DynamicWeightFactor, spec *workv1alpha2.ResourceBindingSpec) ([]workv1alpha2.TargetCluster, error) {
+	switch dynamicWeight {
+	case policyv1alpha1.DynamicWeightByAvailableReplicas:
+		return divideReplicasByResource(clusters, spec, policyv1alpha1.ReplicaDivisionPreferenceWeighted)
+	default:
+		return nil, fmt.Errorf("undefined replica dynamic weight factor: %s", dynamicWeight)
+	}
+}
+
+func divideReplicasByResource(clusters []*clusterv1alpha1.Cluster, spec *workv1alpha2.ResourceBindingSpec,
+	preference policyv1alpha1.ReplicaDivisionPreference, preUsedClustersName ...string) ([]workv1alpha2.TargetCluster, error) {
+	clusterAvailableReplicas := calAvailableReplicas(clusters, spec)
+	clustersMaxReplicas := util.GetSumOfReplicas(clusterAvailableReplicas)
+	if clustersMaxReplicas < spec.Replicas {
+		return nil, fmt.Errorf("clusters resources are not enough to schedule, max %d replicas are support", clustersMaxReplicas)
+	}
```

Comment:
The function `divideReplicasByPreference()`'s parameter `clustersMaxReplicas` is redundant with `clusterAvailableReplicas`. I can tell that you want to save the effort to count the replicas again. 

Can we move the check `if clustersMaxReplicas < spec.Replicas` into `divideReplicasByPreference()`?

Reply:
Cool